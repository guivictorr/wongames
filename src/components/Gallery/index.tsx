import { useEffect, useRef, useState } from 'react'
import SlickSlider, { Settings } from 'react-slick'
import { ArrowBackIos as ArrowLeft } from '@styled-icons/material-outlined/ArrowBackIos'
import { ArrowForwardIos as ArrowRight } from '@styled-icons/material-outlined/ArrowForwardIos'
import Slider from 'components/Slider'

import * as S from './styles'
import { Close } from 'styled-icons/material-outlined'

export type GalleryImageProps = {
  src: string
  label: string
}

export type GalleryProps = {
  items: GalleryImageProps[]
}

const commonSettings: Settings = {
  arrows: true,
  infinite: false,
  lazyLoad: 'ondemand',
  nextArrow: <ArrowRight aria-label="next image" />,
  prevArrow: <ArrowLeft aria-label="previous image" />
}

const settings: Settings = {
  ...commonSettings,
  slidesToShow: 4,
  responsive: [
    {
      breakpoint: 1375,
      settings: {
        arrows: false,
        slidesToShow: 3.2,
        draggable: true
      }
    },
    {
      breakpoint: 1024,
      settings: {
        arrows: false,
        slidesToShow: 2.2,
        draggable: true
      }
    },
    {
      breakpoint: 768,
      settings: {
        arrows: false,
        slidesToShow: 2.2,
        draggable: true
      }
    }
  ]
}

const modalSettings: Settings = {
  ...commonSettings,
  slidesToShow: 1
}

const Gallery = ({ items, ...rest }: GalleryProps) => {
  const [isModalOpen, setIsModalOpen] = useState(false)
  const modalSliderRef = useRef<SlickSlider>(null)

  useEffect(() => {
    const handleKeyUp = ({ key }: KeyboardEvent) => {
      key === 'Escape' && setIsModalOpen(false)
    }

    window.addEventListener('keyup', handleKeyUp)

    return () => window.removeEventListener('keyup', handleKeyUp)
  }, [])

  return (
    <S.Wrapper {...rest} data-cy="gallery">
      <Slider settings={settings} ref={modalSliderRef}>
        {items.map(({ label, src }, index) => (
          <S.Image
            role="button"
            key={`thumb-${label}`}
            src={src}
            alt={`Thumb - ${label}`}
            onClick={() => {
              setIsModalOpen(true)
              modalSliderRef.current?.slickGoTo(index, true)
            }}
          />
        ))}
      </Slider>
      <S.Modal
        aria-label="modal"
        aria-hidden={isModalOpen}
        isOpen={isModalOpen}
      >
        <S.Close
          role="button"
          aria-label="close modal"
          onClick={() => setIsModalOpen(false)}
        >
          <Close size={40} />
        </S.Close>

        <S.Content>
          <Slider ref={modalSliderRef} settings={modalSettings}>
            {items.map(({ label, src }, index) => (
              <S.Image
                role="img"
                key={`gallery-${index}`}
                src={src}
                alt={label}
              />
            ))}
          </Slider>
        </S.Content>
      </S.Modal>
    </S.Wrapper>
  )
}

export default Gallery
