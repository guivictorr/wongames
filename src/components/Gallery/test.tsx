import 'match-media-mock'
import { fireEvent, screen } from '@testing-library/react'
import mockItems from './mock'

import Gallery from '.'
import { renderWithTheme } from 'utils/tests/helpers'

describe('<Gallery />', () => {
  it('should render thumbnails as buttons', () => {
    renderWithTheme(<Gallery items={mockItems.slice(0, 2)} />)

    expect(
      screen.getByRole('button', { name: /thumb - gallery image 1/i })
    ).toHaveAttribute('src', mockItems[0].src)
  })

  it('should open/close modal', () => {
    renderWithTheme(<Gallery items={mockItems.slice(0, 2)} />)

    const modal = screen.getByLabelText('modal')
    expect(modal.getAttribute('aria-hidden')).toBe('false')
    expect(modal).toHaveStyle({
      opacity: 0
    })

    fireEvent.click(
      screen.getByRole('button', { name: /thumb - gallery image 1/i })
    )
    expect(modal.getAttribute('aria-hidden')).toBe('true')
    expect(modal).toHaveStyle({
      opacity: 1
    })
  })

  it('should handle close modal when overlay or button clicked', () => {
    renderWithTheme(<Gallery items={mockItems.slice(0, 2)} />)

    const modal = screen.getByLabelText('modal')

    fireEvent.click(
      screen.getByRole('button', { name: /thumb - gallery image 1/i })
    )

    fireEvent.click(screen.getByLabelText(/close modal/i))
    expect(modal.getAttribute('aria-hidden')).toBe('false')
    expect(modal).toHaveStyle({
      opacity: 0
    })
  })

  it('should handle close modal when click esc', () => {
    const { container } = renderWithTheme(
      <Gallery items={mockItems.slice(0, 2)} />
    )

    const modal = screen.getByLabelText('modal')

    fireEvent.click(
      screen.getByRole('button', { name: /thumb - gallery image 1/i })
    )

    fireEvent.keyUp(container, { key: 'Escape' })
    expect(modal.getAttribute('aria-hidden')).toBe('false')
    expect(modal).toHaveStyle({
      opacity: 0
    })
  })
})
