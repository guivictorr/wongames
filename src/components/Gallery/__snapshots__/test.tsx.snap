// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gallery /> should render thumbnails as buttons 1`] = `
.c4 {
  display: inline-block;
  vertical-align: middle;
  overflow: hidden;
}

.c1 .slick-slider {
  position: relative;
  display: block;
  box-sizing: border-box;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-touch-callout: none;
  -khtml-user-select: none;
  -ms-touch-action: pan-y;
  touch-action: pan-y;
  -webkit-tap-highlight-color: transparent;
}

.c1 .slick-list {
  position: relative;
  display: block;
  overflow: hidden;
  margin: 0;
  padding: 0;
}

.c1 .slick-list:focus {
  outline: none;
}

.c1 .slick-list.dragging {
  cursor: pointer;
  cursor: hand;
}

.c1 .slick-slider .slick-track,
.c1 .slick-slider .slick-list {
  -webkit-transform: translate3d(0,0,0);
  -moz-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  -o-transform: translate3d(0,0,0);
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
}

.c1 .slick-track {
  position: relative;
  top: 0;
  left: 0;
  display: block;
}

.c1 .slick-track:before,
.c1 .slick-track:after {
  display: table;
  content: '';
}

.c1 .slick-track:after {
  clear: both;
}

.c1 .slick-loading .slick-track {
  visibility: hidden;
}

.c1 .slick-slide {
  display: none;
  float: left;
  height: 100%;
  min-height: 1px;
}

.c1 [dir='rtl'] .slick-slide {
  float: right;
}

.c1 .slick-slide img {
  display: block;
}

.c1 .slick-slide.slick-loading img {
  display: none;
}

.c1 .slick-slide.dragging img {
  pointer-events: none;
}

.c1 .slick-initialized .slick-slide {
  display: block;
}

.c1 .slick-loading .slick-slide {
  visibility: hidden;
}

.c1 .slick-vertical .slick-slide {
  display: block;
  height: auto;
  border: 1px solid transparent;
}

.c1 .slick-arrow.slick-hidden {
  display: none;
}

.c0 .slick-prev,
.c0 .slick-next {
  display: block;
  color: #FAFAFA;
  cursor: pointer;
  position: absolute;
  top: 50%;
  width: 2.5rem;
  height: 2.5rem;
  padding: 0;
  -webkit-transform: translate(0,-50%);
  -ms-transform: translate(0,-50%);
  transform: translate(0,-50%);
}

.c0 .slick-prev {
  left: -5.6rem;
}

.c0 .slick-next {
  right: -5.6rem;
}

.c0 .slick-prev.slick-disabled,
.c0 .slick-next.slick-disabled {
  visibility: hidden;
}

.c0 .slick-slide > div {
  margin: 0 1.6rem;
  cursor: pointer;
}

.c0 .slick-list {
  margin: 0 -1.6rem;
}

.c2 {
  opacity: 0;
  pointer-events: none;
  position: fixed;
  width: 100%;
  height: 100%;
  top: 0;
  left: 0;
  background: rgba(0,0,0,0.7);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  z-index: 40;
  -webkit-transition: opacity 0.3s ease-in-out;
  transition: opacity 0.3s ease-in-out;
}

.c3 {
  color: #FAFAFA;
  position: absolute;
  left: 0;
  top: 0;
  cursor: pointer;
  width: 100%;
  height: 100%;
  text-align: right;
}

.c5 {
  max-width: min(120rem,100%);
  max-height: 80rem;
}

@media (max-width:1440px) {
  .c0 {
    overflow-x: hidden;
  }
}

<div
  class="c0"
>
  <section
    class="c1"
  >
    <div
      class="slick-slider slick-initialized"
    >
      
      <div
        class="slick-list"
      >
        <div
          class="slick-track"
          style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
        >
          <div
            aria-hidden="false"
            class="slick-slide slick-active slick-current"
            data-index="0"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <img
                alt="Thumb - Gallery Image 1"
                class=""
                role="button"
                src="/img/games/cyberpunk-1.jpg"
                style="width: 100%; display: inline-block;"
                tabindex="-1"
              />
            </div>
          </div>
          <div
            aria-hidden="false"
            class="slick-slide slick-active"
            data-index="1"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <img
                alt="Thumb - Gallery Image 2"
                class=""
                role="button"
                src="/img/games/cyberpunk-2.jpg"
                style="width: 100%; display: inline-block;"
                tabindex="-1"
              />
            </div>
          </div>
        </div>
      </div>
      
      
    </div>
  </section>
  <div
    aria-hidden="false"
    aria-label="modal"
    class="c2"
  >
    <div
      aria-label="close modal"
      class="c3"
      role="button"
    >
      <svg
        aria-hidden="true"
        class="c4"
        fill="currentColor"
        focusable="false"
        height="40"
        viewBox="0 0 24 24"
        width="40"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 0h24v24H0V0z"
          fill="none"
        />
        <path
          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"
        />
      </svg>
    </div>
    <div
      class="c5"
    >
      <section
        class="c1"
      >
        <div
          class="slick-slider slick-initialized"
          dir="ltr"
        >
          <svg
            aria-hidden="true"
            aria-label="previous image"
            class="c4 slick-arrow slick-prev slick-disabled"
            data-role="none"
            fill="currentColor"
            focusable="false"
            style="display: block;"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0V0z"
              fill="none"
              opacity="0.87"
            />
            <path
              d="M17.51 3.87L15.73 2.1 5.84 12l9.9 9.9 1.77-1.77L9.38 12l8.13-8.13z"
            />
          </svg>
          <div
            class="slick-list"
          >
            <div
              class="slick-track"
              style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
            >
              <div
                aria-hidden="false"
                class="slick-slide slick-active slick-current"
                data-index="0"
                style="outline: none; width: 0px;"
                tabindex="-1"
              >
                <div>
                  <img
                    alt="Gallery Image 1"
                    class=""
                    role="img"
                    src="/img/games/cyberpunk-1.jpg"
                    style="width: 100%; display: inline-block;"
                    tabindex="-1"
                  />
                </div>
              </div>
              <div
                aria-hidden="true"
                class="slick-slide"
                data-index="1"
                style="outline: none; width: 0px;"
                tabindex="-1"
              />
            </div>
          </div>
          <svg
            aria-hidden="true"
            aria-label="next image"
            class="c4 slick-arrow slick-next"
            data-role="none"
            fill="currentColor"
            focusable="false"
            style="display: block;"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M24 24H0V0h24v24z"
              fill="none"
              opacity="0.87"
            />
            <path
              d="M6.49 20.13l1.77 1.77 9.9-9.9-9.9-9.9-1.77 1.77L14.62 12l-8.13 8.13z"
            />
          </svg>
        </div>
      </section>
    </div>
  </div>
</div>
`
